File: server/models/multiplayer.py
"""多人联机模型 — Room / RoomConfig / PlayerSession / CharacterSlot 等"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime


# ─── 房间配置 ───

class RoomConfig(BaseModel):
    """房间配置参数（房主创建时设定，游戏开始后不可修改）"""

    # 基本设置
    room_name: str = Field(default="冒险小队", min_length=1, max_length=30)
    max_players: int = Field(default=4, ge=1, le=10)
    password: Optional[str] = None  # 可选房间密码，空=公开房间

    # 故事设置
    storybook_id: Optional[str] = None  # None=使用服务器默认故事书
    starting_region: Optional[str] = None  # None=由GM决定
    narrative_style: str = "detailed"  # concise | detailed | epic | humorous
    narrative_language: str = "zh"  # zh | en | auto

    # 游戏规则
    difficulty_modifier: int = Field(default=0, ge=-5, le=5)
    allow_npc_recruitment: bool = True
    enable_fast_travel: bool = True
    death_penalty: str = "soft"  # permanent | soft | narrative_only
    rest_recovery_multiplier: float = Field(default=1.0, ge=0.5, le=2.0)

    # 行动轮规则
    action_round_timeout: int = Field(default=300, ge=60, le=600)
    auto_skip_on_timeout: bool = True
    allow_skip_action: bool = True

    # 房间管理
    allow_late_join: bool = True
    late_join_intro_delay: int = Field(default=2, ge=1, le=5)
    allow_spectators: bool = False


# ─── 玩家会话 ───

class PlayerSession(BaseModel):
    """房间中的玩家会话"""
    player_id: str
    player_name: str = ""
    character_id: Optional[str] = None
    character_name: Optional[str] = None
    character_background: Optional[str] = None  # GM引入用
    is_host: bool = False
    is_ready: bool = False
    is_online: bool = True
    last_heartbeat: str = Field(default_factory=lambda: datetime.now().isoformat())
    status: str = "waiting"  # waiting | creating_character | ready | in_game | spectating | pending_intro | disconnected
    slot_id: Optional[str] = None  # inherit模式关联的角色槽
    joined_at_round: int = 0  # 加入时的轮数


# ─── 角色槽（继承存档模式） ───

class CharacterSlot(BaseModel):
    """继承存档时的角色槽（只读，等待玩家认领）"""
    slot_id: str
    character_id: str
    character_name: str
    character_summary: str = ""  # "人类战士 Lv.3"
    claimed_by_player_id: Optional[str] = None


# ─── 房间状态 ───

class RoomState(BaseModel):
    """房间的游戏状态"""
    phase: str = "waiting"  # waiting | preparing | playing | ended
    world_day: int = 1
    current_round: int = 0
    location: Optional[Dict[str, Any]] = None
    players_acted: List[str] = []
    round_start_time: Optional[str] = None
    common_backstory: Optional[str] = None  # GM生成的统一起始故事


# ─── 房间 ───

class Room(BaseModel):
    """多人房间"""
    room_id: str
    host_player_id: str
    config: RoomConfig
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    started_at: Optional[str] = None
    ended_at: Optional[str] = None

    players: List[PlayerSession] = []
    state: RoomState = Field(default_factory=RoomState)
    mode: str = "new"  # new | inherit
    character_slots: List[CharacterSlot] = []

    current_round_actions: Dict[str, str] = {}
    current_round_dice_results: Dict[str, Any] = {}
    last_round_result: Optional[Dict[str, Any]] = None
    narrative_history: List[Dict[str, Any]] = []
    room_notifications: List[Dict[str, Any]] = []


# ─── 请求/响应模型 ───

class CreateRoomRequest(BaseModel):
    mode: str = "new"  # new | inherit
    config: RoomConfig
    inherit_data: Optional[Dict[str, Any]] = None


class JoinRoomRequest(BaseModel):
    password: Optional[str] = None
    claimed_slot_id: Optional[str] = None  # inherit模式指定认领的角色槽


class ClaimSlotRequest(BaseModel):
    slot_id: str


class CharacterReadyRequest(BaseModel):
    character_id: str
    character_name: str
    character_data: Optional[Dict[str, Any]] = None
    character_background: Optional[str] = None


class StartGameRequest(BaseModel):
    starting_location: Optional[Dict[str, Any]] = None
    common_backstory_narrative: Optional[str] = None


class SubmitActionRequest(BaseModel):
    action: str = Field(min_length=1, max_length=2000)
    dice_result: Optional[Dict[str, Any]] = None


class RoomResponse(BaseModel):
    room_id: str
    host_player_id: str
    config: RoomConfig
    mode: str
    created_at: str
    started_at: Optional[str] = None
    state: RoomState
    players: List[PlayerSession]
    character_slots: List[CharacterSlot] = []


class RoundStatusResponse(BaseModel):
    current_round: int
    players_acted: List[str]
    pending_players: List[str]
    actions: Dict[str, str]
    dice_results: Dict[str, Any] = {}
    round_start_time: Optional[str] = None
    timeout_at: Optional[str] = None
    latest_round_result: Optional[Dict[str, Any]] = None


class RoundProcessResponse(BaseModel):
    round: int
    player_actions: Dict[str, str]
    dice_results: Dict[str, Any] = {}
    conflicts: List[Dict[str, Any]] = []
    narrative: str
    consequences: Dict[str, Any] = {}
    world_state_changes: Dict[str, Any] = {}
    introduced_players: List[Dict[str, Any]] = []
    next_round: int


class SaveResponse(BaseModel):
    """房主请求存档时，服务器返回的打包数据"""
    archive_id: str
    archive_name: str
    created_at: str
    world_day: int
    current_round: int
    location: Optional[Dict[str, Any]] = None
    player_characters: Dict[str, Any] = {}
    shared_world_state: Optional[Dict[str, Any]] = None
    chronicle_entries: List[Dict[str, Any]] = []
    player_list: List[Dict[str, Any]] = []


class CommonBackstoryResponse(BaseModel):
    common_backstory: str
    suggested_starting_location: Optional[Dict[str, Any]] = None
    individual_hooks: Dict[str, str] = {}


class NarrativeHistoryResponse(BaseModel):
    round: int
    narrative: str
    player_actions: Dict[str, str]
    dice_results: Dict[str, Any] = {}
    timestamp: str


class RoomNotification(BaseModel):
    event: str  # spectator_joined | character_created | player_introduced
    player_id: str
    player_name: str = ""
    character_name: str = ""
    character_background: str = ""
    round: int = 0
    narrative: str = ""
    timestamp: str = ""


class RoomNotificationsResponse(BaseModel):
    notifications: List[Dict[str, Any]] = []

File: client/src/types/multiplayer.ts
/**
 * 多人联机系统类型定义
 * 对应服务端 models/multiplayer.py
 */

import type { Character } from './character';

// ─── 房间配置 ───

export type NarrativeStyle = 'concise' | 'detailed' | 'epic' | 'humorous';
export type NarrativeLanguage = 'zh' | 'en' | 'auto';
export type DeathPenalty = 'permanent' | 'soft' | 'narrative_only';

export const NARRATIVE_STYLE_LABELS: Record<NarrativeStyle, string> = {
  concise: '简洁',
  detailed: '细腻',
  epic: '宏大',
  humorous: '幽默',
};

export const DEATH_PENALTY_LABELS: Record<DeathPenalty, string> = {
  permanent: '永久死亡',
  soft: '重伤可恢复',
  narrative_only: '仅叙事',
};

export interface RoomConfig {
  // 基本设置
  roomName: string;           // 1-30字
  maxPlayers: number;         // 1-10
  password?: string;          // 可选密码

  // 故事设置
  storybookId?: string;       // null=服务器默认
  startingRegion?: string;    // null=GM决定
  narrativeStyle: NarrativeStyle;
  narrativeLanguage: NarrativeLanguage;

  // 游戏规则
  difficultyModifier: number; // -5 ~ +5
  allowNpcRecruitment: boolean;
  enableFastTravel: boolean;
  deathPenalty: DeathPenalty;
  restRecoveryMultiplier: number; // 0.5 ~ 2.0

  // 行动轮规则
  actionRoundTimeout: number; // 60 ~ 600 秒
  autoSkipOnTimeout: boolean;
  allowSkipAction: boolean;

  // 房间管理
  allowLateJoin: boolean;
  lateJoinIntroDelay: number;  // 1 ~ 5 轮
  allowSpectators: boolean;
}

export const DEFAULT_ROOM_CONFIG: RoomConfig = {
  roomName: '冒险小队',
  maxPlayers: 4,
  password: undefined,
  storybookId: undefined,
  startingRegion: undefined,
  narrativeStyle: 'detailed',
  narrativeLanguage: 'zh',
  difficultyModifier: 0,
  allowNpcRecruitment: true,
  enableFastTravel: true,
  deathPenalty: 'soft',
  restRecoveryMultiplier: 1.0,
  actionRoundTimeout: 300,
  autoSkipOnTimeout: true,
  allowSkipAction: true,
  allowLateJoin: true,
  lateJoinIntroDelay: 2,
  allowSpectators: false,
};

export const ROOM_PRESETS: Record<string, Partial<RoomConfig>> = {
  beginner: {
    roomName: '新手冒险',
    difficultyModifier: 0,
    narrativeStyle: 'detailed',
    actionRoundTimeout: 300,
    deathPenalty: 'soft',
    allowNpcRecruitment: true,
  },
  hardcore: {
    roomName: '硬核征途',
    difficultyModifier: 3,
    narrativeStyle: 'detailed',
    actionRoundTimeout: 180,
    deathPenalty: 'permanent',
    allowNpcRecruitment: true,
  },
  casual: {
    roomName: '轻松时光',
    difficultyModifier: -2,
    narrativeStyle: 'humorous',
    actionRoundTimeout: 600,
    deathPenalty: 'narrative_only',
    allowNpcRecruitment: true,
  },
  speedrun: {
    roomName: '速通模式',
    difficultyModifier: 0,
    narrativeStyle: 'concise',
    actionRoundTimeout: 120,
    deathPenalty: 'soft',
    allowNpcRecruitment: false,
  },
  epic: {
    roomName: '史诗之旅',
    difficultyModifier: 1,
    narrativeStyle: 'epic',
    actionRoundTimeout: 600,
    deathPenalty: 'soft',
    allowNpcRecruitment: true,
  },
};

export const PRESET_LABELS: Record<string, string> = {
  beginner: '新手友好',
  hardcore: '硬核写实',
  casual: '轻松幽默',
  speedrun: '速通快节奏',
  epic: '史诗长篇',
};

// ─── 玩家会话 ───

export type PlayerStatus =
  | 'waiting'
  | 'creating_character'
  | 'ready'
  | 'in_game'
  | 'spectating'
  | 'pending_intro'
  | 'disconnected';

export const PLAYER_STATUS_LABELS: Record<PlayerStatus, string> = {
  waiting: '等待中',
  creating_character: '创建角色中',
  ready: '已就绪',
  in_game: '游戏中',
  spectating: '观战中',
  pending_intro: '等待引入',
  disconnected: '已离线',
};

export interface RoomPlayer {
  playerId: string;
  playerName: string;
  characterId: string | null;
  characterName: string | null;
  characterBackground: string | null;
  isHost: boolean;
  isReady: boolean;
  isOnline: boolean;
  lastHeartbeat: string;
  status: PlayerStatus;
  slotId?: string;
  joinedAtRound: number;
}

// ─── 角色槽（继承存档模式） ───

export interface CharacterSlot {
  slotId: string;
  characterId: string;
  characterName: string;
  characterSummary: string;
  claimedByPlayerId: string | null;
}

// ─── 房间状态 ───

export type RoomPhase = 'waiting' | 'preparing' | 'playing' | 'ended';

export interface RoomLocation {
  region: string;
  subRegion: string;
  specificPlace: string;
  coordinates: { x: number; y: number; z: number };
}

export interface RoomState {
  phase: RoomPhase;
  worldDay: number;
  currentRound: number;
  location: RoomLocation | null;
  playersActed: string[];
  roundStartTime: string | null;
  commonBackstory: string | null;
}

// ─── 房间 ───

export type RoomMode = 'new' | 'inherit';

export interface Room {
  roomId: string;
  hostPlayerId: string;
  config: RoomConfig;
  mode: RoomMode;
  createdAt: string;
  startedAt: string | null;
  state: RoomState;
  players: RoomPlayer[];
  characterSlots: CharacterSlot[];
  roomNotifications?: RoomNotification[];
}

// ─── 多人存档 ───

export interface MultiplayerSaveData {
  archiveId: string;
  archiveName: string;
  createdAt: string;
  roomId: string;
  worldDay: number;
  currentRound: number;
  location: RoomLocation | null;
  playerCharacters: Record<string, Character>;
  sharedWorldState: Record<string, unknown>;
  chronicleEntries: MultiplayerChronicleEntry[];
  playerList: Array<{ playerId: string; characterName: string }>;
}

// ─── 行动轮 ───

export interface RoundStatus {
  currentRound: number;
  playersActed: string[];
  pendingPlayers: string[];
  actions: Record<string, string>;
  diceResults: Record<string, unknown>;
  roundStartTime: string | null;
  timeoutAt: string | null;
  latestRoundResult?: RoundResult | null;
  recentNotifications?: RoomNotification[];
}

export interface RoomNotification {
  event: 'spectator_joined' | 'character_created' | 'player_introduced';
  playerId: string;
  playerName: string;
  characterName: string;
  characterBackground: string;
  round: number;
  narrative: string;
  timestamp: string;
}

export interface RoomNotificationsResponse {
  notifications: RoomNotification[];
}

export interface RoundSubmission {
  playersActed: string[];
  totalPlayers: number;
  isRoundComplete: boolean;
  roundResult?: RoundResult;
}

export interface SpectatorReadyResponse {
  status: string;
  message: string;
  estimatedIntroRound: number;
}

export interface Conflict {
  type: 'target_conflict' | 'space_conflict' | 'causal_conflict';
  players: string[];
  target?: string;
  resolution: 'simultaneous' | 'sequential';
  description?: string;
}

export interface RoundResult {
  round: number;
  playerActions: Record<string, string>;
  diceResults: Record<string, unknown>;
  conflicts: Conflict[];
  narrative: string;
  consequences: Record<string, unknown>;
  worldStateChanges: Record<string, unknown>;
  introducedPlayers: Array<{
    playerId: string;
    characterName: string;
    narrative: string;
  }>;
  nextRound: number;
}

// ─── 编年史 ───

export interface MultiplayerChronicleEntry {
  id: string;
  worldDay: number;
  round: number;
  location: {
    region: string;
    subRegion: string;
    specificPlace: string;
  };
  playerActions: Record<string, {
    action: string;
    diceResult?: unknown;
    skipped: boolean;
  }>;
  conflicts: Conflict[];
  narrative: string;
  playerConsequences: Record<string, unknown>;
  worldStateChanges: Record<string, unknown>;
  timestamp: string;
}

// ─── 共同背景故事 ───

export interface CommonBackstory {
  commonBackstory: string;
  suggestedStartingLocation: RoomLocation | null;
  individualHooks: Record<string, string>;
}

// ─── 叙事历史 ───

export interface NarrativeHistory {
  round: number;
  narrative: string;
  playerActions: Record<string, string>;
  diceResults: Record<string, unknown>;
  timestamp: string;
}

// ─── 单人实时同步 ───

export interface NearbyPlayer {
  playerId: string;
  characterName: string;
  region: string;
  subRegion: string;
  coordinates: { x: number; y: number; z: number };
  currentAction: string;
  status: string;
  worldDay: number;
  lastHeartbeat: string;
}

export interface RealtimeSessionUpload {
  characterName: string;
  region: string;
  subRegion: string;
  coordinates: { x: number; y: number; z: number };
  worldDay: number;
  currentAction: string;
  status: string;
}
